documentation/private-link-service/private-link-redirect-map.md

```diff
@@ -0,0 +1,258 @@
+   | entry_attr.SAI_OUTBOUND_CA_TO_PA_ENTRY_ATTR_SERVICE_REWRITE_SIP_MASK | `sai_ip_address_t` | `ffff:ffff:ffff:ffff:ffff:ffff::` |
+   | entry_attr.SAI_OUTBOUND_CA_TO_PA_ENTRY_ATTR_SERVICE_REWRITE_DIP | `sai_ip_address_t` | `2603:10e1:100:2::` |
+   | entry_attr.SAI_OUTBOUND_CA_TO_PA_ENTRY_ATTR_SERVICE_REWRITE_DIP_MASK | `sai_ip_address_t` | `ffff:ffff:ffff:ffff:ffff:ffff::` |
+
+The service rewrite info and the port map id could be cached in pipeline/packet metadata and later retrieved in the stage of outbound port map.
+
+**Mapping - outbound port map**:
+
+In this stage, it accomplishes the operation of PL redirect map with the following steps:
+
+1. It validates the port map object exists according to the port map id. If the port map is missing, eni drop counter `SAI_ENI_STAT_OUTBOUND_PORT_MAP_MISS_DROP_PACKETS` increases with one and the pipeline drops the packet, otherwise continues to the next.
+
+1. It looks up the table `outbound port range` with the match key (port_map_id, packet.overlay.destination_port). If none of table entry is matched, it drops the packet and increases eni drop counter `SAI_ENI_STAT_OUTBOUND_PORT_RANGE_ENTRY_MISS_DROP_PACKETS` with 1, otherwise continues to the next.
+
+1. One table entry is matched. If the entry attribute ACTION is `skip_mapping`, the operation in this stage will be skipped, otherwise goes to the next.
+
+1. The entry attribute ACTION must be `map_to_private_link_service`, it starts to do PL redirect map with the entry attr info and the cached service rewrite info. Assume the matched entry attributes are like below:
+
+   | SAI field name | Type | Value |
+   | --- | --- | --- |
+   | entry.outbound_port_map_id | `sai_object_id_t` | (SAI object ID of the port map) |
+   | entry.dst_port_range | `sai_u32_range_t` | 2000-2100 |
+   | entry_attr.SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_ACTION | `sai_outbound_port_map_port_range_entry_action_t` | `SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ACTION_MAP_TO_PRIVATE_LINK_SERVICE` |
+   | entry_attr.SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_BACKEND_IP | `sai_ip_address_t` | `3.3.3.1` |
+   | entry_attr.SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_MATCH_PORT_BASE | `sai_uint16_t` | `2000` |
+   | entry_attr.SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_BACKEND_PORT_BASE | `sai_uint16_t` | `11000` |
+
+    And packet overlay destination port is 2010, the operation details are as below:
+
+    - port translation
+
+    ```c
+    packet.overlay.destination_port = 11000 + 2010 - 2000
+    ```
+
+    - nat 4to6
+
+    ```c
+    packet.overlay.ipv6.src_addr = ((( (IPv6Address)packet.overlay.ipv4.src_addr & ~svc_rewrite_info.src_prefix_mask) | svc_rewrite_info.src_prefix_addr) & ~eni.attr_SAI_ENI_ATTR_PL_SIP_MASK) | eni.attr_SAI_ENI_ATTR_PL_SIP
+    packet.overlay.ipv6.dst_addr = (svc_rewrite_info.dst_prefix_addr & ~((bit<128>)0xFFFFFFFF)) | (bit<128>)'3.3.3.1'
```

Comment:
svc_rewrite_info.dst_prefix_mask should be used.

Reply:
updated